reps/2022-04-21-object_store_HA.md

```diff
@@ -0,0 +1,197 @@
+
+**Pros**:
+
+- Less intrusive to the design of object store.
+
+**Cons**:
+
+- Deployment is more complicated.
+- The modification of the reference table is a high-frequency operation. Syncing with an external storage may hurt performance.
+- Potentially some tricky consistency issues when an owner fails.
+
+##### Option 2: Rebuild failed global owner via the information in the whole cluster
+
+There are two options, the active way and passive way, to rebuild the reference table and the RPC connections when a global owner restarts.
+
+**The Active Way**:
+
+The global owner actively collect information about the alive objects it owns to rebuild the reference table. Here are the details steps of global owner `G` rebuild:
+
+1. `G` begins to rebuild the reference table, and sets the rebuilding status to `REBUILDING`.
+2. `G` sends RPCs to all Raylets and ask every Raylet to traverse all local workers and reply with the information of all objects owned by `G`. `G` then uses the information to rebuild the reference table and re-establish RPC connections.
+3. `G` sets the rebuilding state to `READY`.
+
+**Note**:
+
+- RPCs sent by other workers to `G` should be retried until the state of `G's` Actor becomes `Alive`.
+- The reply callback of some RPCs will not be invoked until the rebuilding state of `G` is set to `READY`.
+
+**The Passive Way**:
+
+Every Raylet maintains an RPC connection to every global owner to watch the status of the global owners. When the global owner `G` dies and the job is not finished yet. Each raylet will collect and merge the reference table of all workers on this node, and report to `G`. Details:
+
+1. Raylet will find out that `G` is dead through RPC disconnection.
+2. When Raylet knows that `G` is restarted, Raylet sends the information below to `G`:
+    - References of all objects which are owned by `G`. (Collected by traversing all local workers.)
+    - Objects which are owned by `G` and are local (the data is in the local Object Store).
+    - The spill URLs of locally spilled objects which are owned by `G`.
+3. `G` will reconnect to the Raylet and the workers on it after receiving the above collected information from this Raylet.
+4. `G` sets the state to `READY` after finished rebuilding reference table.
+5. The reply callback of some RPCs (`WaitForObjectFree`) will not be called until the rebuilding state of `G` is set to `READY`.
+
+**Note**:
+
+- RPCs sent by other workers to `G` should be retried until the state of `G's` Actor becomes `Alive`.
+- The reply callback of some RPCs will not be invoked until the rebuilding state of `G` is set to `READY`.
+
+We prefer __Option.2__, and due to the difficulty of implementation, we are more trend to choose the __active__ way to restore `G`
```

Comment:
I'll add some notes here for the cons of the passive way.

* Raylets need to subscribe to the actor state notifications in order to know the new RPC address of a restarted actor and reconnect to it. To implement this, we need to add something similar to "actor manager" in core worker to Raylet. This is an additional coding effort.
* Raylets pushing collected information to the restarted actor v.s. the actor pulling information from Raylets is just like push-based v.s. pull-based resource reporting. This is what we've already discussed and we've concluded that pull-based resource reporting is easier to maintain and test. I believe this conclusion also applies to object reference and location reporting.

Reply:
I will add it to the doc's body.